### ISSUE TITLE
[TASK-207] Implement KB Gap Detector Agent

### ISSUE THREAD

**Issue by OthmanMohammad** [OthmanMohammad/multi-agent-support-system#95]
# [TASK-207] Implement KB Gap Detector Agent

**Task ID:** TASK-207
**Story:** [STORY-002] Knowledge Base Swarm
**Epic:** [EPIC-001] Tier 1: Essential Foundation
**Status:** TODO
**Priority:** P1
**Effort:** 1 day
**Labels:** `task`, `tier-1`, `knowledge-base`, `p1`, `agent`

---

## 📋 Overview

Implement the KB Gap Detector agent that identifies missing content in the knowledge base by analyzing queries with low KB match scores and clustering similar unanswered questions.

---

## 🎯 Objective

Create an agent that:
1. Tracks queries with low KB match scores (<0.7)
2. Clusters similar queries using embeddings
3. Identifies common topics without good articles
4. Suggests new article topics
5. Prioritizes gaps by frequency and impact

---

## 📂 File Location

```
src/agents/essential/knowledge_base/gap_detector.py
```

---

### Class Structure

```python
from typing import List, Dict
from collections import Counter
from sentence_transformers import SentenceTransformer
from sklearn.cluster import DBSCAN
import numpy as np

from src.agents.base.base_agent import BaseAgent
from src.workflow.state import AgentState
from src.database.models import Message, Conversation
from src.database.connection import get_db_session

class KBGapDetector(BaseAgent):
    """
    KB Gap Detector Agent

    Identifies missing knowledge base content.
    """

    def __init__(self):
        super().__init__(
            agent_type="kb_gap_detector",
            model="claude-3-haiku-20240307",
            temperature=0.2
        )
        self.db_session = get_db_session()
        self.embedding_model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
        print("✓ KB Gap Detector initialized")

    def detect_gaps(self, days: int = 30, min_frequency: int = 5) -> List[Dict]:
        """
        Detect KB content gaps

        Args:
            days: Look back this many days
            min_frequency: Min occurrences to consider

        Returns:
            List of gap topics with priority
        """
        # Get queries with low KB scores
        low_match_queries = self._get_low_match_queries(days)

        if len(low_match_queries) < min_frequency:
            return []

        # Cluster similar queries
        clusters = self._cluster_queries(low_match_queries)

        # Analyze each cluster
        gaps = []
        for cluster_id, queries in clusters.items():
            if len(queries) >= min_frequency:
                gap = self._analyze_cluster(queries)
                gaps.append(gap)

        # Sort by priority
        gaps.sort(key=lambda x: x['priority_score'], reverse=True)

        return gaps

    def _get_low_match_queries(self, days: int) -> List[str]:
        """Get queries with low KB match scores"""
        from datetime import datetime, timedelta

        cutoff_date = datetime.utcnow() - timedelta(days=days)

        # Query conversations with low KB scores
        conversations = self.db_session.query(Conversation).filter(
            Conversation.created_at >= cutoff_date,
            Conversation.kb_match_score < 0.7  # Low match
        ).all()

        queries = []
        for conv in conversations:
            # Get first user message
            first_msg = self.db_session.query(Message).filter(
                Message.conversation_id == conv.id,
                Message.role == 'user'
            ).first()

            if first_msg:
                queries.append(first_msg.content)

        return queries

    def _cluster_queries(self, queries: List[str]) -> Dict[int, List[str]]:
        """Cluster similar queries using embeddings"""
        if not queries:
            return {}

        # Generate embeddings
        embeddings = self.embedding_model.encode(queries)

        # Cluster using DBSCAN
        clustering = DBSCAN(eps=0.3, min_samples=3, metric='cosine').fit(embeddings)

        # Group queries by cluster
        clusters = {}
        for idx, label in enumerate(clustering.labels_):
            if label == -1:  # Noise
                continue

            if label not in clusters:
                clusters[label] = []

            clusters[label].append(queries[idx])

        return clusters

    def _analyze_cluster(self, queries: List[str]) -> Dict:
        """Analyze cluster and generate gap report"""
        # Use LLM to summarize common topic
        system_prompt = """Analyze these related customer queries and identify the common topic.

Return JSON:
{
  "topic": "Brief topic description",
  "suggested_article_title": "Proposed KB article title",
  "key_questions": ["Q1", "Q2", "Q3"],
  "category": "billing|technical|usage|api|integrations"
}"""

        user_prompt = f"""Queries:\n""" + "\n".join(f"- {q}" for q in queries[:10])

        try:
            import json
            response = self.call_llm(system_prompt, user_prompt, max_tokens=512)
            analysis = json.loads(response)
        except:
            # Fallback
            analysis = {
                "topic": "Unknown topic",
                "suggested_article_title": "New Article Needed",
                "key_questions": queries[:3],
                "category": "general"
            }

        # Calculate priority score
        priority_score = self._calculate_priority(queries, analysis)

        return {
            "topic": analysis["topic"],
            "suggested_article_title": analysis["suggested_article_title"],
            "key_questions": analysis.get("key_questions", []),
            "category": analysis.get("category", "general"),
            "frequency": len(queries),
            "priority_score": priority_score,
            "example_queries": queries[:5]
        }

    def _calculate_priority(self, queries: List[str], analysis: Dict) -> float:
        """Calculate gap priority score (0-100)"""
        # Frequency component (50%)
        frequency_score = min(len(queries) * 5, 50)

        # Category importance (30%)
        category_weights = {
            "billing": 30,
            "technical": 25,
            "api": 25,
            "usage": 20,
            "integrations": 20,
            "general": 10
        }
        category_score = category_weights.get(analysis.get("category", "general"), 10)

        # Recency (20%) - more recent queries score higher
        recency_score = 20  # Simplified

        return frequency_score + category_score + recency_score
```

---

## ✅ Acceptance Criteria

- [ ] Detects queries with low KB match scores
- [ ] Clusters similar queries using DBSCAN
- [ ] Generates gap reports with suggested topics
- [ ] Prioritizes gaps by frequency and category
- [ ] Returns top N gaps
- [ ] Unit tests pass

---

## 🧪 Test Cases

```python
def test_cluster_similar_queries(gap_detector):
    """Test clustering of similar queries"""
    queries = [
        "How do I export data?",
        "Can I export my data?",
        "Data export instructions?",
        "How to integrate with Slack?",
        "Slack integration help"
    ]

    clusters = gap_detector._cluster_queries(queries)

    # Should have 2 clusters (export + slack)
    assert len(clusters) >= 1

def test_detect_gaps(gap_detector, db_session):
    """Test gap detection"""
    gaps = gap_detector.detect_gaps(days=30, min_frequency=3)

    assert isinstance(gaps, list)
    if gaps:
        assert "topic" in gaps[0]
        assert "priority_score" in gaps[0]
```

---

**Created:** 2025-11-15
**Assignee:** @OthmanMohammad
